### PR TITLE
Basic Scaffold To Build and Test

### PR DESCRIPTION
**Pull request by emileswarts**
Add Docker and Docker Compose
Add Makefile and Rspec
Install Sinatra

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+	$(MAKE) serve
+	./bin/wait_for_mysql
+	docker-compose run --rm app rspec
+	$(MAKE) stop
+
+stop:
+	docker-compose kill
+	docker-compose rm -f
+
+.PHONY: test serve
```
Comment (RyanMacG) [inline Makefile]: Should add the stop command here too

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,16 @@
+serve:
+	$(MAKE) stop
+	docker-compose build
+	docker-compose up -d
+
+test:
+	$(MAKE) serve
```
Comment (adrianclay) [inline Makefile]: Do we need to serve the app locally to run the tests? Wouldn’t it be quicker to bring up the db by itself?

Comment (adrianclay) [reply Makefile]: Unless we were thinking end to end tests?

File: mysql/Dockerfile [hunk 3]
```diff
@@ -0,0 +1 @@
+FROM mysql:5.7
```
Comment (adrianclay) [inline mysql/Dockerfile]: If we don’t have a special MySQL image, be easier to use the image name from the `docker-compose.yml`

File: Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,64 @@
+pipeline {
+  agent any
```
Comment (adrianclay) [inline Jenkinsfile]: You’re going to run into that port conflict issue again without the concurrent executor fix, unless you’ve fixed that elsewhere.

File: Gemfile [hunk 5]
```diff
@@ -0,0 +1,10 @@
+source 'http://rubygems.org'
+ruby File.read('.ruby-version').chomp
+
+gem 'sinatra'
+gem 'puma'
+gem 'mysql2'
+
+group :test do
+  gem 'rspec'
```
Comment (adrianclay) [inline Gemfile]: The lock file is suspiciously empty...

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,10 @@
+FROM ruby:2.5
+
+WORKDIR /usr/src/app
+
+COPY Gemfile Gemfile.lock .ruby-version ./
+RUN bundle check || bundle install --jobs 20 --retry 5
```
Comment (adrianclay) [inline Dockerfile]: I’m guessing the lock file is empty because you run bundle inside of docker.

File: Makefile [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+	./bin/wait_for_mysql
+	docker-compose run --rm app rspec
```
Comment (adrianclay) [inline Makefile]: Be nice to have an RSpec file committed to see it running the tests. No biggie though.

File: Jenkinsfile (hunk 4 above)
Comment (emileswarts) [reply Jenkinsfile]: We've set the jenkins to only run a single execution at a time for now

File: Jenkinsfile [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+      }
+    }
+
+    stage('Deploy') {
+      when{
+        branch 'master'
+      }
+
+      steps {
+        deploy()
```
[the remark this answers is not in the document]
Comment (emileswarts) [reply Jenkinsfile]: This is because I changed 'master' to this branch just to see it working.
Will be run exclusively on master from now on.

**Pull request merged**